### ISSUE TITLE
:unused-value (to detect unused transient / pure fn ret vals)

### ISSUE THREAD

**Issue by slipset** [clj-kondo/clj-kondo#1258]
Some of the functions on transients, specifically `assoc!` appears to be working even when not handling the return type:
```Clojure
user> (let [t (transient {})]
        (assoc! t :foo :bar)
        (persistent! t))
;; => {:foo :bar}
user>
```
This blows up when we reach more than 8 entries in the map:
```Clojure
user> (let [t (transient {})]
        (doseq [i (range 10)]
          (assoc! t (str i) i))
        (persistent! t))
;; => {"0" 0, "1" 1, "2" 2, "3" 3, "4" 4, "5" 5, "6" 6, "7" 7}
user> 
```

`conj!` does not exhibit such behaviour, but I believe that's an implementation detail, so a linter should most likely deal with all the operations on transients.

**Describe the solution you'd like**
I would like a warning, much akin to the `discarded constants` linter, so any time an operation on a transient is not in a returned position (if that's even a word), it should be flagged.

**Describe alternatives you've considered**
No real alternatives other than doing nothing, which is perfectly valid

With some guiding, I'd be happy to take a stab at a linter like this.

**Comment by borkdude** [comment]
I'm not entirely sure how to tackle this one reliably but there is already a `:redundant-expression` linter which may be good to look at first for inspiration.

**Comment by slipset** [comment]
From Eastwood: 
https://github.com/jonase/eastwood/blob/master/README.md#unused-ret-vals

**Comment by borkdude** [comment]
Thanks. Looks like it more generically applicable to pure functions too. I'll rename this issue similarly.
I think we can just use the name from eastwood for this linter (but make it singular like most other names in clj-kondo): `:unused-ret-val`.

There is a prototype of this linter in `warn-unused-transient` which is only used for `assoc!` and `conj` at the moment, but we can probably generalize this the same way eastwood did it.

**Comment by esuomi** [comment]
Bringing a semi-old related discussion from Clojurians here, fashionably late:

Consider this code snippet:
```clojure
(defn do-lots-of-things
  [xs]
  (when (a-thing?)
    (for [x xs] (do-more-things x)))
  (when (b-thing?)
    (do-more-stuff)))
```

The use of `for` comprehension here causes its body to not get evaluated because it is not in a tail position. Tail position analysis is already done for `recur` so something similar should be possible for this case as well, and would help tons in highlighting the misuse.

Original thread: https://clojurians.slack.com/archives/CHY97NXE2/p1649145002582929

**Comment by borkdude** [comment]
Don't forget to use a thumbs up if you're interested in seeing this feature implemented.

**Comment by borkdude** [comment]
<img width="726" alt="Screenshot 2022-09-18 at 13 58 33" src="https://user-images.githubusercontent.com/284934/190901841-04c962eb-831f-4dac-9c3e-81de1a2d105d.png">

**Comment by borkdude** [comment]
Still have to support the `doseq` case.

**Comment by borkdude** [comment]
The example from @esuomi is supported but in a more limited fashion, e.g. 

```
(defn foo [] (map inc [1 2 3]) nil)
```

will trigger a warning.

Will get back to the more advanced case once this linter sees more real life usage.
